flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,431 @@
+		 */
+		private long currentReceivedCheckpointId = -1L;
+
+		/** The number of opened channels. */
+		private int numOpenChannels;
+
+		private final ChannelStateWriter channelStateWriter;
+
+		private final CheckpointBarrierUnaligner handler;
+
+		public ThreadSafeUnaligner(
+				int totalNumChannels,
+				ChannelStateWriter channelStateWriter,
+				CheckpointBarrierUnaligner handler) {
+			storeNewBuffers = new boolean[totalNumChannels];
+			this.channelStateWriter = channelStateWriter;
+			this.handler = handler;
+			numOpenChannels = totalNumChannels;
+		}
+
+		@Override
+		public synchronized void notifyBarrierReceived(CheckpointBarrier barrier, InputChannelInfo channelInfo) {
+			long barrierId = barrier.getId();
+
+			if (currentReceivedCheckpointId < barrierId) {
+				handleNewCheckpoint(barrier);
+				handler.runInTaskThread(() -> handler.notifyCheckpoint(barrier, 0, 0), "notifyCheckpoint");
+			}
+
+			int channelIndex = handler.getFlattenedChannelIndex(channelInfo);
+			if (barrierId == currentReceivedCheckpointId && storeNewBuffers[channelIndex]) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("{}: Received barrier from channel {} @ {}.", handler.taskName, channelIndex, barrierId);
+				}
+
+				storeNewBuffers[channelIndex] = false;
+
+				if (++numBarriersReceived == numOpenChannels) {
+					allBarriersReceivedFuture.complete(null);
+				}
+			}
+		}
+
+		@Override
+		public synchronized void notifyBufferReceived(Buffer buffer, InputChannelInfo channelInfo) {
+			if (storeNewBuffers[handler.getFlattenedChannelIndex(channelInfo)]) {
+				channelStateWriter.addInputData(
+					currentReceivedCheckpointId,
+					channelInfo,
+					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+					buffer);
+			} else {
+				buffer.recycleBuffer();
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+			allBarriersReceivedFuture.cancel(false);
+		}
+
+		private synchronized void handleNewCheckpoint(CheckpointBarrier barrier) {
+			long barrierId = barrier.getId();
+			if (!allBarriersReceivedFuture.isDone()) {
+				// we did not complete the current checkpoint, another started before
+				LOG.warn("{}: Received checkpoint barrier for checkpoint {} before completing current checkpoint {}. " +
+						"Skipping current checkpoint.",
+					handler.taskName,
+					barrierId,
+					currentReceivedCheckpointId);
+
+				// let the task know we are not completing this
+				long currentCheckpointId = currentReceivedCheckpointId;
+				handler.runInTaskThread(() ->
+					handler.notifyAbort(currentCheckpointId,
+						new CheckpointException(
+							"Barrier id: " + barrierId,
+							CheckpointFailureReason.CHECKPOINT_DECLINED_SUBSUMED)),
+					"notifyAbort");
+			}
+
+			currentReceivedCheckpointId = barrierId;
+			Arrays.fill(storeNewBuffers, true);
+			numBarriersReceived = 0;
+			allBarriersReceivedFuture = new CompletableFuture<>();
+			channelStateWriter.start(barrierId, barrier.getCheckpointOptions());
```

Comment:
As I understood, `channelStateWriter.finishInput` is called when this `allBarriersReceivedFuture` is completed. 
But in a quite entangled manner: `StreamTask.prepareInputSnapshot -> InputProcessor -> InputGate -> Unaligner -> ThreadSafeUnaligner (this`).
Why don't we set this callback right here?

Reply:
`finishInput` is called when all gates and their `Unaligner` have triggered `allBarriersReceivedFuture`. Within one `Unaligner` its very hard to decide if indeed `finishInput` should be called.
In particular, `StreamTwoInputProcessor#prepareSnapshot` and `StreamMultipleInputProcessor#prepareSnapshot` are combining the futures of the gates/unaligners to joint futures and I have no idea on how to map it differently.